src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,50 +1,209 @@
+        nodeByIndex.prev.next = newNode;
+        nodeByIndex.prev = newNode;
+        size++;
+    }
 
+    private void addAtBeginning(T value) {
+        Node<T> tempNode = new Node<>(null, value, head);
+        head.prev = tempNode;
+        head = tempNode;
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T listItem: list) {
+            add(listItem);
+        }
+
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIsIndexExist(index, size - 1, "Invalid index");
+
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIsIndexExist(index, size - 1, "Invalid index");
+
+        Node<T> nodeByIndex = findNodeByIndex(index);
+        T valueBeforeSet = nodeByIndex.item;
+        nodeByIndex.item = value;
+
+        return valueBeforeSet;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIsIndexExist(index, size - 1, "Invalid index");
+        return unlink(findNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> headNode = head;
+        if (Objects.equals(head.item, object)) {
+            removeFirstElement();
+            return true;
+        }
+
+        while (headNode != null) {
+            if (headNode.item == null || headNode.item.equals(object)) {
+                removeFromMiddle(headNode);
+                return true;
+            }
+            headNode = headNode.next;
+        }
+
         return false;
     }
 
+    private void removeFirstElement() {
+        Node<T> newHead = head.next;
+        if (newHead != null) {
+            newHead.prev = null;
+            head = newHead;
+        } else {
+            head = null;
+        }
+        size--;
+    }
+
+    private void removeFromMiddle(Node<T> nodeByIndex) {
+        Node<T> previous = nodeByIndex.prev;
+        Node<T> next = nodeByIndex.next;
+        previous.next = next;
+        next.prev = previous;
+        size--;
+    }
+
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T unlink(Node<T> targetNode) {
+        final T element = targetNode.item;
+        Node<T> prev = targetNode.prev;
+        Node<T> next = targetNode.next;
+
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            targetNode.prev = null;
+        }
+
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            targetNode.next = null;
+        }
+
+        size--;
+        return element;
+    }
+
+    private void checkIsIndexExist(int index, int size, String message) {
```

Comment:
if the message is the same everywhere, maybe we should not pass it?
```suggestion
    private void checkIfIndexExists(int index, int size) { // or just `checkIndex()`
```